operator/main.go

```diff
@@ -0,0 +1,165 @@
+	"github.com/gravitational/trace"
+	//+kubebuilder:scaffold:imports
+)
+
+var (
+	scheme   = runtime.NewScheme()
+	setupLog = ctrl.Log.WithName("setup")
+
+	//go:embed config/crd/bases/*.teleport.dev_*.yaml
+	crdFS embed.FS
+)
+
+func init() {
+	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
+
+	utilruntime.Must(resourcesv5.AddToScheme(scheme))
+	utilruntime.Must(resourcesv2.AddToScheme(scheme))
+	//+kubebuilder:scaffold:scheme
+
+	utilruntime.Must(apiextv1.AddToScheme(scheme))
+}
+
+func main() {
+	ctx := ctrl.SetupSignalHandler()
+
+	var err error
+	var metricsAddr string
+	var probeAddr string
+	flag.StringVar(&metricsAddr, "metrics-bind-address", ":8080", "The address the metric endpoint binds to.")
+	flag.StringVar(&probeAddr, "health-probe-bind-address", ":8081", "The address the probe endpoint binds to.")
+	opts := zap.Options{
+		Development: true,
+	}
+	opts.BindFlags(flag.CommandLine)
+	flag.Parse()
+
+	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
+
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		Scheme:                 scheme,
+		MetricsBindAddress:     metricsAddr,
+		Port:                   9443,
+		HealthProbeBindAddress: probeAddr,
+		LeaderElection:         true,
+		LeaderElectionID:       "431e83f4.teleport.dev",
+	})
+	if err != nil {
+		setupLog.Error(err, "unable to start manager")
+		os.Exit(1)
+	}
+
+	var teleportClient *client.Client
+
+	retry, err := utils.NewLinear(utils.LinearConfig{
+		Step: 100 * time.Millisecond,
+		Max:  time.Second,
+	})
+	if err != nil {
+		setupLog.Error(err, "failed to setup retry")
+		os.Exit(1)
+	}
+	if err := retry.For(ctx, func() error {
+		teleportClient, err = sidecar.NewSidecarClient(ctx, sidecar.Options{})
+		if err != nil {
+			setupLog.Error(err, "failed to connect to teleport cluster, backing off")
+		}
+		return err
+	}); err != nil {
+		setupLog.Error(err, "failed to setup teleport client")
+		os.Exit(1)
+	}
+	setupLog.Info("connected to Teleport")
+
+	if err = (&resourcescontrollers.RoleReconciler{
+		Client:         mgr.GetClient(),
+		Scheme:         mgr.GetScheme(),
+		TeleportClient: teleportClient,
+	}).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", "Role")
+		os.Exit(1)
+	}
+	if err = (&resourcescontrollers.UserReconciler{
+		Client:         mgr.GetClient(),
+		Scheme:         mgr.GetScheme(),
+		TeleportClient: teleportClient,
+	}).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", "User")
+		os.Exit(1)
+	}
+	//+kubebuilder:scaffold:builder
+
+	if err := mgr.AddHealthzCheck("healthz", healthz.Ping); err != nil {
+		setupLog.Error(err, "unable to set up health check")
+		os.Exit(1)
+	}
+	if err := mgr.AddReadyzCheck("readyz", healthz.Ping); err != nil {
+		setupLog.Error(err, "unable to set up ready check")
+		os.Exit(1)
+	}
+
+	// Install CRDs
+	if err := mgr.Add(manager.RunnableFunc(func(ctx context.Context) error {
```

Comment:
Have we confirmed that this isn't racy with the Controller watches/caches being established ?

Reply:
I tried this multiple times and didn't found an issue, but can't guarantee it won't :sweat_smile: 

I added a `WaitForCacheSync` before installing the CRDs, it should prevent races in case they are possible. Better be safe

Reply:
Ah yeah, I should have expanded, but my concern was that I couldn't recall how the kube api handles you trying to establish a watch on a resource that hasn't yet been defined. In previous operators I have built, we have had the user manually deploy the CRDs prior to deploying the operator, I definitely prefer this model of the operator deploying it's own CRDs though.

Reply:
Perfect, those logs confirm that we handle this situation. I think the retry negates the need for some additional complexity here.